backend/howtheyvote/api/query.py

```diff
@@ -254,3 +298,96 @@ def query(self, query: str | None = None) -> Self:
+        else:
+            query = XapianQuery(
+                XapianQuery.OP_AND_MAYBE,
+                query,
+                self._xapian_featured_subquery(),
+            )
+
+            query = XapianQuery(
+                XapianQuery.OP_AND_MAYBE,
+                query,
+                self._xapian_age_subquery(),
+            )
+
+            query = XapianQuery(
+                XapianQuery.OP_AND_MAYBE,
+                query,
+                self._xapian_phrase_subquery(index),
+            )
+
+        for field, value in self.get_filters().items():
+            # Fields have to be indexed as boolean terms to be used as filters
+            term = boolean_term(field, value)
+            query = XapianQuery(XapianQuery.OP_FILTER, query, XapianQuery(term))
+
+        return query
+
+    def _xapian_phrase_subquery(self, index: Database) -> XapianQuery:
+        # This is a phrase subquery, i.e. it matches documents that contain the terms of the
+        # search query in the original order. It's used to boost phrase matches even if
+        # a user hasn't explicitly specified a phrase query.
+        parser = self._xapian_query_parser(index)
+        parser.set_default_op(XapianQuery.OP_PHRASE)
+        query = parser.parse_query(self.get_query())
+
+        return XapianQuery(
+            XapianQuery.OP_SCALE_WEIGHT,
+            query,
+            self.BOOST_PHRASE,
+        )
+
+    def _xapian_featured_subquery(self) -> XapianQuery:
+        # This subquery assigns a constant weight to featured votes and 0 otherwise.
+        return XapianQuery(
+            XapianQuery.OP_SCALE_WEIGHT,
+            XapianQuery(ValueWeightPostingSource(field_to_slot("is_featured"))),
+            self.BOOST_FEATURED,
+        )
+
+    def _xapian_age_subquery(self) -> XapianQuery:
+        # This subquery assigns a decreasing weight based on age, i.e. recent votes get a
+        # higher weight.
+        now = datetime.datetime.now().timestamp()
+        max_diff = datetime.timedelta(days=self.AGE_DECAY_DAYS).total_seconds()
+
+        age_source = ValueDecayWeightPostingSource(field_to_slot("timestamp"))
+        age_source.set_max_diff(max_diff)
+        age_source.set_origin(now)
+
+        return XapianQuery(
+            XapianQuery.OP_SCALE_WEIGHT,
+            XapianQuery(age_source),
+            self.BOOST_AGE,
+        )
+
+    def _xapian_bm25_weight(self) -> Weight:
+        # This parameter controls document length normalization. As we currently index mostly
+        # short vote titles and a few keywords, we set b=0 to disable normalization. We do
+        # not want a vote to score higher just because its title is shorter. If that changes
+        # (and we start indexing longer documents), we should reconsider this.
+        b = 0  # Xapian default: 0.5
+
+        # The following BM25 parameters are Xapian’s defaults.
+        # See: https://xapian.org/docs/apidoc/html/classXapian_1_1BM25Weight.html
+        k1 = 1
```

Comment:
Maybe adjust this to give less weight to votes just because they contain a search term multiple times. For example, when searching for "Ukraine", a vote that contains Ukraine in the title and as a geographic area will always be ranked higher than a vote that contains it only in the title, even if the latter is more recent. Not sure this is justified.

Reply:
This sounds reasonable, maybe we should start with `k1 = 0 and go from there, as your argument makes sense?